In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split

<h1> Wczytanie danych <h1>

In [17]:
data=pd.read_csv("data.csv")
data.head(5)

Unnamed: 0  pclass  survived  \
0           0       1         1   
1           1       1         1   
2           2       1         0   
3           3       1         0   
4           4       1         0   

                                              name     sex      age  sibsp  \
0                    Allen, Miss. Elisabeth Walton  female  29.0000      0   
1                   Allison, Master. Hudson Trevor    male   0.9167      1   
2                     Allison, Miss. Helen Loraine  female   2.0000      1   
3             Allison, Mr. Hudson Joshua Creighton    male  30.0000      1   
4  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female  25.0000      1   

   parch  ticket      fare    cabin embarked boat   body  \
0      0   24160  211.3375       B5        S    2    NaN   
1      2  113781  151.5500  C22 C26        S   11    NaN   
2      2  113781  151.5500  C22 C26        S  NaN    NaN   
3      2  113781  151.5500  C22 C26        S  NaN  135.0   
4      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  KolumnNull CabinReduced  
0                     St Louis, MO           0            B  
1  Montreal, PQ / Chesterville, ON           0            C  
2  Montreal, PQ / Chesterville, ON           1            C  
3  Montreal, PQ / Chesterville, ON           1            C  
4  Montreal, PQ / Chesterville, ON           1            C

Funkcja train_test_split przyjmuje parametry:
- test_size
- train_size (oba określają rozmiar danych testowych/treningowych w stosunku do cąłości zbioru, jeżeli jeden tylko zostanie ukreślony, drugi jest dopełnieniem do całości; wartość z przedziału 0-1)
- arrays - zbiór danych (dataframe, numpy array, lista itd)
- random_state - int odpowiadający za poziom przetasowania danych przed dokonaniem podziału na zbiory

Wartość zwracana to dwa obiekty - zbiór testowy i zbiór treningowy

In [18]:
col_name = ['sex','cabin','CabinReduced']
data[col_name]

sex    cabin CabinReduced
0     female       B5            B
1       male  C22 C26            C
2     female  C22 C26            C
3       male  C22 C26            C
4     female  C22 C26            C
...      ...      ...          ...
1304  female      NaN          NaN
1305  female      NaN          NaN
1306    male      NaN          NaN
1307    male      NaN          NaN
1308    male      NaN          NaN

[1309 rows x 3 columns]

<h1> Podział zbioru na zbiór treningowy i zbiór testowy <h1>

In [6]:
X_train, X_test, y_train, y_test = train_test_split(data[col_name],data['survived'],test_size=0.30,random_state = 0)

In [5]:
print(X_train.shape,X_test.shape,y_train.shape,y_test.shape)


(916, 3) (393, 3) (916,) (393,)


X_test i y_test liczą po 393 obserwacji, co stanowi 30% całego zbioru danych (co zgadza się z tym co zostało podane w poleceniu funkcji train_test_split)


<h1>Kardynalność zmiennych<h1>

In [6]:
Unique_test = [x for x in X_test.sex.unique() if x not in X_train.sex.unique()]
len(Unique_test)

0

In [7]:
Unique_test = [x for x in X_test.cabin.unique() if x not in X_train.cabin.unique()]
len(Unique_test)

37

In [8]:
Unique_test = [x for x in X_test.CabinReduced.unique() if x not in X_train.CabinReduced.unique()]
len(Unique_test)

0

Rozkład cech w zależności od kardynalności dla zmiennej sex i cabin jest równomierny, natomiast dla CabinReduced nie

<h1> Kodowanie zmiennych kategorycznych do zmiennych liczbowych <h1>

In [19]:
Dict_cabin_reduced={}
for i,x in enumerate(data['CabinReduced'].unique()):
    Dict_cabin_reduced[x]=i
Dict_cabin_reduced

{'B': 0, 'C': 1, 'E': 2, 'D': 3, 'A': 4, nan: 5, 'T': 6, 'F': 7, 'G': 8}

In [20]:
Dict_cabin={}
for i,x in enumerate(data['cabin'].unique()):
    Dict_cabin[x]=i
Dict_cabin

{'B5': 0,
 'C22 C26': 1,
 'E12': 2,
 'D7': 3,
 'A36': 4,
 'C101': 5,
 nan: 6,
 'C62 C64': 7,
 'B35': 8,
 'A23': 9,
 'B58 B60': 10,
 'D15': 11,
 'C6': 12,
 'D35': 13,
 'C148': 14,
 'C97': 15,
 'B49': 16,
 'C99': 17,
 'C52': 18,
 'T': 19,
 'A31': 20,
 'C7': 21,
 'C103': 22,
 'D22': 23,
 'E33': 24,
 'A21': 25,
 'B10': 26,
 'B4': 27,
 'E40': 28,
 'B38': 29,
 'E24': 30,
 'B51 B53 B55': 31,
 'B96 B98': 32,
 'C46': 33,
 'E31': 34,
 'E8': 35,
 'B61': 36,
 'B77': 37,
 'A9': 38,
 'C89': 39,
 'A14': 40,
 'E58': 41,
 'E49': 42,
 'E52': 43,
 'E45': 44,
 'B22': 45,
 'B26': 46,
 'C85': 47,
 'E17': 48,
 'B71': 49,
 'B20': 50,
 'A34': 51,
 'C86': 52,
 'A16': 53,
 'A20': 54,
 'A18': 55,
 'C54': 56,
 'C45': 57,
 'D20': 58,
 'A29': 59,
 'C95': 60,
 'E25': 61,
 'C111': 62,
 'C23 C25 C27': 63,
 'E36': 64,
 'D34': 65,
 'D40': 66,
 'B39': 67,
 'B41': 68,
 'B102': 69,
 'C123': 70,
 'E63': 71,
 'C130': 72,
 'B86': 73,
 'C92': 74,
 'A5': 75,
 'C51': 76,
 'B42': 77,
 'C91': 78,
 'C125': 79,
 'D10 D12': 80,
 'B82 

In [21]:
Dict_sex={}
for i, x in enumerate(data.sex.unique()):
    Dict_sex[x]=i
Dict_sex

{'female': 0, 'male': 1}

<h1> Mapowanie zmienych <h1>

In [22]:
X_test['cabin_map']=X_test['cabin'].map(Dict_cabin)
X_train['cabin_map']=X_train['cabin'].map(Dict_cabin)
X_train['cabin_map_reduced']=X_train['CabinReduced'].map(Dict_cabin_reduced)
X_test['cabin_map_reduced']=X_test['CabinReduced'].map(Dict_cabin_reduced)
X_test['sex_map']=X_test['sex'].map(Dict_sex)
X_train['sex_map']=X_train['sex'].map(Dict_sex)


In [23]:
print(f'Nulle dla zbioru testowego :\n{pd.isnull(X_test).sum()}\n')
print(f'Nulle dla zbioru treningowego :\n{pd.isnull(X_train).sum()}')


Nulle dla zbioru testowego :
sex                    0
cabin                312
CabinReduced           0
cabin_map              0
cabin_map_reduced    312
sex_map                0
dtype: int64

Nulle dla zbioru treningowego :
sex                    0
cabin                702
CabinReduced           0
cabin_map              0
cabin_map_reduced    702
sex_map                0
dtype: int64


<p> W moim przypadku brakujące wartości dla zmiennych powstałych z mapowania już są ustawione jako wartość zero, dlatego w powyższym podsumowaniu nie ma wartości NaN dla nowych zmiennych. <p>

<h1> Podsumowanie <h1>

In [24]:
print('Liczba etykiet zmiennej CabinReduced test: {}'.format(len(X_test["CabinReduced"].unique())))
print('Liczba etykiet zmiennej CabinReduced train: {}'.format(len(X_train["CabinReduced"].unique())))
print('Liczba etykiet zmiennej cabin_map_reduced test: {}'.format(len(X_test["cabin_map_reduced"].unique())))
print('Liczba etykiet zmiennej cabin_map_reduced train: {}'.format(len(X_train["cabin_map_reduced"].unique())))


Liczba etykiet zmiennej CabinReduced test: 8
Liczba etykiet zmiennej CabinReduced train: 9
Liczba etykiet zmiennej cabin_map_reduced test: 8
Liczba etykiet zmiennej cabin_map_reduced train: 9


In [25]:
print('Liczba etykiet zmiennej cabin test: {}'.format(len(X_test["cabin"].unique())))
print('Liczba etykiet zmiennej cabin train: {}'.format(len(X_train["cabin"].unique())))
print('Liczba etykiet zmiennej cabin_map test: {}'.format(len(X_test["cabin_map"].unique())))
print('Liczba etykiet zmiennej cabin_map train: {}'.format(len(X_train["cabin_map"].unique())))


Liczba etykiet zmiennej cabin test: 71
Liczba etykiet zmiennej cabin train: 151
Liczba etykiet zmiennej cabin_map test: 71
Liczba etykiet zmiennej cabin_map train: 151


In [26]:
print('Liczba etykiet zmiennej sex test: {}'.format(len(X_test["sex"].unique())))
print('Liczba etykiet zmiennej sex train: {}'.format(len(X_train["sex"].unique())))
print('Liczba etykiet zmiennej sex_map test: {}'.format(len(X_test["sex_map"].unique())))
print('Liczba etykiet zmiennej sex_map train: {}'.format(len(X_train["sex_map"].unique())))


Liczba etykiet zmiennej sex test: 2
Liczba etykiet zmiennej sex train: 2
Liczba etykiet zmiennej sex_map test: 2
Liczba etykiet zmiennej sex_map train: 2


<p> Ogólnie zbiór treningowy ma większą liczebność niż zbiór testowy, ponieważ jest znacznie większy (stosunek 30% do 70%).  Liczebność etykiet zmiennej cabin uległa znacznej redukcji. Z 151 do 8 dla zbioru treningowego i z 71 do 8 dla zbioru testowego. Proces redukcji etykiet i zamiany zmiennych kategorycznych na zmienne liczbowe ma duże znaczenie dla tworzenia modelu i predykcji. Zmniejszenie liczebności etykiet chroni przed overfittingiem i dominacją zmiennej o dużej kardynalności nad zmiennymi o mniejszej kardynalności. Natomiast zmiana wartości na liczbowe upraszcza tworzenie modelu.<p>